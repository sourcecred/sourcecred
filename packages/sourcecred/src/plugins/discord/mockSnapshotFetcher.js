// @flow

import path from "path";
import fs from "fs-extra";
import {Fetcher} from "./fetcher";
import {createHash} from "crypto";

async function snapshotFetch(url: string | Request | URL): Promise<Response> {
  const snapshotDir = "src/plugins/discord/snapshots";
  const shasum = createHash("sha256");
  shasum.update(Buffer.from(((url: any): string), "utf8"));
  const filename = shasum.digest("hex");
  const file = path.join(snapshotDir, filename);
  if (await fs.exists(file)) {
    const contents = await fs.readFile(file);
    return new Response(contents, {status: 200});
  } else {
    throw new Error(`couldn't load snapshot for ${file} (${String(url)})`);
  }
}

export const snapshotFetcher = (): Fetcher =>
  new Fetcher({
    fetch: snapshotFetch,
    token: "mock-token",
  });
