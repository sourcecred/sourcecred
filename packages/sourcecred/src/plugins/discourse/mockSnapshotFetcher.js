// @flow

import deepFreeze from "deep-freeze";
import {createHash} from "crypto";
import path from "path";
import fs from "fs-extra";
import {Fetcher, type DiscourseFetchOptions} from "./fetch";

export const options: DiscourseFetchOptions = deepFreeze({
  serverUrl: "https://sourcecred-test.discourse.group",
});

async function snapshotFetch(url: string | Request | URL): Promise<Response> {
  const snapshotDir = "src/plugins/discourse/snapshots";
  const shasum = createHash("sha256");
  shasum.update(Buffer.from(((url: any): string), "utf8"));
  const filename = shasum.digest("hex");
  const file = path.join(snapshotDir, filename);
  if (await fs.exists(file)) {
    const contents = await fs.readFile(file);
    return new Response(contents, {status: 200});
  } else {
    throw new Error(`couldn't load snapshot for ${file}`);
  }
}
export const snapshotFetcher = (): Fetcher =>
  new Fetcher(options, snapshotFetch, 0);
