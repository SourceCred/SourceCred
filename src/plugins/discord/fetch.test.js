// @flow

import {snapshotFetcher} from "./mockSnapshotFetcher";

describe("plugins/discord/fetch", () => {
  describe("snapshot testing", () => {
    it("loads the snapshots", async () => {
      expect(await snapshotFetcher().guilds()).toMatchSnapshot();
    });
  });
});
