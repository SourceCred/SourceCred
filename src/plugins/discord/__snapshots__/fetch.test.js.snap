// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/discord/fetch snapshot testing loads the snapshots 1`] = `
Array [
  Object {
    "id": "678348980639498428",
    "name": "SourceCred Test Server",
    "permissions": 33620992,
  },
]
`;
